### ISSUE TITLE
Error in message_digest and optional URL encoding/decoding

### ISSUE THREAD

**Issue by vvanelslande** [yhirose/cpp-httplib#289]
I'm using the latest version of Visual Studio Build Tools 2019.
The error is a C2131 here https://github.com/yhirose/cpp-httplib/blob/72b20c08da8fa70050f510c199bf6b0e94a2b532/httplib.h#L2288
I also need a way to disable URL encoding/decoding to use my own URL encoding/decoding code.

**Comment by vvanelslande** [comment]
you didn't add optional URL encoding/decoding

**Comment by yhirose** [comment]
@vvanelslande, thanks for the good catch! I fixed it.

When it comes to "I also need a way to disable URL encoding/decoding to use my own URL encoding/decoding code.", it's a separate issue.  cpp-httplib right now doesn't allow us to use own URL encoding/decoding code. I just wonder why you need to use your own version of URL encoding/decoding code instead of the build-in in cpp-httplib? Is the build-in one missing something that your version has?

**Comment by vvanelslande** [comment]
I want to use this library unmodified in my [Citra (Nintendo 3DS emulator) Canary fork](https://github.com/vvanelslande/citra).

in citra-emu/citra#4754, the PR's author says that they had to edit the URL encoding.

**Comment by vvanelslande** [comment]
I added a ifndef to skip defining `encode_url` and `decode_url`.

```cpp
#ifndef HTTPLIB_CUSTOM_URL_ENCODE_DECODE
inline std::string encode_url(const std::string &s) {
  std::string result;

  for (auto i = 0; s[i]; i++) {
    switch (s[i]) {
    case ' ': result += "%20"; break;
    case '+': result += "%2B"; break;
    case '\r': result += "%0D"; break;
    case '\n': result += "%0A"; break;
    case '\'': result += "%27"; break;
    case ',': result += "%2C"; break;
    case ':': result += "%3A"; break;
    case ';': result += "%3B"; break;
    default:
      auto c = static_cast<uint8_t>(s[i]);
      if (c >= 0x80) {
        result += '%';
        char hex[4];
        size_t len = snprintf(hex, sizeof(hex) - 1, "%02X", c);
        assert(len == 2);
        result.append(hex, len);
      } else {
        result += s[i];
      }
      break;
    }
  }

  return result;
}

inline std::string decode_url(const std::string &s) {
  std::string result;

  for (size_t i = 0; i < s.size(); i++) {
    if (s[i] == '%' && i + 1 < s.size()) {
      if (s[i + 1] == 'u') {
        int val = 0;
        if (from_hex_to_i(s, i + 2, 4, val)) {
          // 4 digits Unicode codes
          char buff[4];
          size_t len = to_utf8(val, buff);
          if (len > 0) { result.append(buff, len); }
          i += 5; // 'u0000'
        } else {
          result += s[i];
        }
      } else {
        int val = 0;
        if (from_hex_to_i(s, i + 1, 2, val)) {
          // 2 digits hex codes
          result += static_cast<char>(val);
          i += 2; // '00'
        } else {
          result += s[i];
        }
      }
    } else if (s[i] == '+') {
      result += ' ';
    } else {
      result += s[i];
    }
  }

  return result;
}
#endif
```

usage in my header:
```cpp
#ifdef HTTPLIB_CUSTOM_URL_ENCODE_DECODE
namespace httplib::detail {
std::string encode_url(const std::string& s);
std::string decode_url(const std::string& s);
} // namespace httplib::detail
#endif

#include <httplib.h>
```

in the cpp file:
```cpp
namespace httplib::detail {

std::string encode_url(const std::string& s) {
    std::string result;
    return result;
}

std::string decode_url(const std::string& s) {
    std::string result;
    return result;
}

} // namespace httplib::detail
```

**Comment by yhirose** [comment]
@vvanelslande, could you make a new issue for the optional URL encoding/decoding, and move the comments to it, since it's a separate issue. Thanks!

**Comment by vvanelslande** [comment]
nvm, I think it was a bug in Citra, the `Request::path` was set to the URL.

**Comment by yhirose** [comment]
@vvanelslande, I actually found that cpp-httplib incorrectly escapes ':' when sending a request to a proxy server.  I fixed it at f2476f2. When you have time, could you try with the latest httplib.h in the master branch? Hope the fix works on your project. I appreciate your report!

**Comment by vvanelslande** [comment]
@yhirose the [current HTTP code](https://github.com/vvanelslande/citra/blob/master/src/core/hle/service/http_c.cpp) crashes and I don't want to try to fix that.
